File: Data/Examples/CSharpTester/MCTester/GUI/ObjectWorld/Scan/ScanPolygonGeometry.cs
using System;
using System.Collections.Generic;
using System.Text;
using MapCore;
using MCTester.Managers.ObjectWorld;
using UnmanagedWrapper;

namespace MCTester.ObjectWorld
{
    public class ScanPolygonGeometry : MCTester.ObjectWorld.ScanGeometryBase,IDNEditModeCallback
    {
        DNSMcScanPolygonGeometry m_ScanPolygonGeometry;
        IDNMcEditMode m_EditMode;
        private DNSQueryParams m_SpatialQueryParams;
        bool m_CompletelyInside;
        DNEMcPointCoordSystem mScanCoordSys;
        private IDNEditModeCallback m_Callback;

        public ScanPolygonGeometry(DNEMcPointCoordSystem coordSys, DNSQueryParams SQParams, bool CompletelyInsideOnly, float Tolerance)
        {
            m_CompletelyInside = CompletelyInsideOnly;
            m_SpatialQueryParams = SQParams;
            m_EditMode = MCTester.Managers.MapWorld.MCTMapFormManager.MapForm.EditMode;
            mScanCoordSys = coordSys;
            m_Callback = null;
        }

        public void StartPolyScan()
        {
            try
            {
                IDNMcOverlay activeOverlay = MCTester.Managers.ObjectWorld.Manager_MCOverlayManager.ActiveOverlay;
                DNSMcVector3D[] locationPoints = new DNSMcVector3D[0];


                IDNMcObjectSchemeItem ObjSchemeItem = DNMcPolygonItem.Create(DNEItemSubTypeFlags._EISTF_SCREEN,
                                                                                DNELineStyle._ELS_SOLID,
                                                                                DNSMcBColor.bcBlackOpaque,
                                                                                2f,
                                                                                null,
																				new DNSMcFVector2D(0, -1),
                                                                                1f,
                                                                                DNEFillStyle._EFS_SOLID,
                                                                                new DNSMcBColor(0,255,100,100));
                    
                IDNMcObject obj = DNMcObject.Create(activeOverlay,
                                                    ObjSchemeItem,
                                                    DNEMcPointCoordSystem._EPCS_SCREEN,
                                                    locationPoints,
                                                    false);

                //In order to prevent retrieval of the polygon in the scan 
                ObjSchemeItem.Detectibility = false;
                ((IDNMcPolygonItem)ObjSchemeItem).SetDrawPriorityGroup(DNEDrawPriorityGroup._EDPG_TOP_MOST, (uint)DNEPredefinedPropertyIDs._EPPI_SHARED_PROPERTY_ID, false);
                ((IDNMcPolygonItem)ObjSchemeItem).SetDrawPriority(SByte.MaxValue, (uint)DNEPredefinedPropertyIDs._EPPI_SHARED_PROPERTY_ID, false); 

                uint[] activeViewportID = new uint[1];
                activeViewportID[0] = MCTester.Managers.MapWorld.MCTMapFormManager.MapForm.Viewport.ViewportID;
                IDNMcConditionalSelector viewportSelector = DNMcViewportConditionalSelector.Create(MCTester.Managers.MapWorld.MCTMapFormManager.MapForm.Viewport.OverlayManager,
                                                                                                        DNEViewportType._EVT_ALL_VIEWPORTS,
                                                                                                        DNEViewportCoordinateSystem._EVCS_ALL_COORDINATE_SYSTEMS,
                                                                                                        activeViewportID,
                                                                                                        true);
                obj.SetConditionalSelector(DNEActionType._EAT_VISIBILITY,
                                            true,
                                            viewportSelector);
                obj.SetDrawPriority(SByte.MaxValue);

                MCTester.Managers.MapWorld.MCTMapFormManager.MapForm.EditMode.StartInitObject(obj, ObjSchemeItem);

                m_Callback = m_EditMode.GetEventsCallback();
                m_EditMode.SetEventsCallback(this);

            }
            catch (MapCoreException McEx)
            {
                MapCore.Common.Utilities.ShowErrorMessage("PolyScan", McEx);
            }
        }


        #region IDNEditModeCallback Members

        public void ExitAction(int nExitCode)
        {
            m_EditMode.SetEventsCallback(m_Callback);
        }

        public void NewVertex(IDNMcObject pObject, IDNMcObjectSchemeItem pItem, DNSMcVector3D WorldVertex, DNSMcVector3D ScreenVertex, uint uVertexIndex, double dAngle)
        {
        }

        public void PointDeleted(IDNMcObject pObject, IDNMcObjectSchemeItem pItem, DNSMcVector3D WorldVertex, DNSMcVector3D ScreenVertex, uint uVertexIndex)
        {
        }

        public void PointNewPos(IDNMcObject pObject, IDNMcObjectSchemeItem pItem, DNSMcVector3D WorldVertex, DNSMcVector3D ScreenVertex, uint uVertexIndex, double dAngle, bool bDownOnHeadPoint)
        {
        }

        public void ActiveIconChanged(IDNMcObject pObject, IDNMcObjectSchemeItem pItem, DNEPermission eIconPermission, uint uIconIndex)
        {
        }

        public void InitItemResults(IDNMcObject pObject, IDNMcObjectSchemeItem pItem, int nExitCode)
        {
            if (nExitCode != 0)
            {
                IDNMcMapViewport CurrViewport = MCTester.Managers.MapWorld.MCTMapFormManager.MapForm.Viewport;

                IDNMcObjectScheme objScheme = pObject.GetScheme();
                IDNMcObjectSchemeNode[] objSchemeNode = objScheme.GetNodes(DNENodeKindFlags._ENKF_ANY_ITEM);

                DNSMcVector3D[] Coords = objSchemeNode[0].GetCoordinates(CurrViewport, /*DNEMcPointCoordSystem._EPCS_SCREEN*/mScanCoordSys, pObject);

                m_ScanPolygonGeometry = new DNSMcScanPolygonGeometry(mScanCoordSys, Coords);
                m_ScanPolygonGeometry.uGeometryType = DNGEOMETRY_TYPE.Polygon;

                //Remove polygon Item from the map
                pObject.Remove();
                objScheme.Dispose();

                DNSTargetFound[] TargetFound = null;
                try
                {
                    TargetFound = CurrViewport.ScanInGeometry(m_ScanPolygonGeometry, m_CompletelyInside, m_SpatialQueryParams);
                    ShowScanResult(m_ScanPolygonGeometry, TargetFound);
                }
                catch (MapCoreException McEx)
                {
                    MapCore.Common.Utilities.ShowErrorMessage("ScanInGeometry", McEx);
                }
            }
        }

        public void EditItemResults(IDNMcObject pObject, IDNMcObjectSchemeItem pItem, int nExitCode)
        {
        }

		public void DragMapResults(IDNMcMapViewport pViewport, DNSMcVector3D NewCenter)
        {
        }

		public void RotateMapResults(IDNMcMapViewport pViewport, float fNewYaw, float fNewPitch)
        {
        }

		public void DistanceDirectionMeasureResults(IDNMcMapViewport pViewport, DNSMcVector3D WorldVertex1, DNSMcVector3D WorldVertex2, double dDistance, double dAngle)
        {
        }

		public void DynamicZoomResults(IDNMcMapViewport pViewport, float fNewScale, DNSMcVector3D NewCenter)
        {
        }

		public void CalculateHeightResults(IDNMcMapViewport pViewport, double dHeight, DNSMcVector3D[] coords, int status)
        {
        }

		public void CalculateVolumeResults(IDNMcMapViewport pViewport, double dVolume, DNSMcVector3D[] coords, int status)
        {
        }

        #endregion
    }
}

File: Data/Examples/Web/web-library/src/services/scan/scanPolygonGeometry.ts
import objectWorldService from "../objectWorld.service";
import { ViewportData } from "../../model/viewportData";
import mapcoreData from "../../mapcore-data";
import { runMapCoreSafely } from "../error-handler.service";
// import scanService from "../../scan.service";

class ScanPolygonGeometry {
    ScanPolygonGeometry: MapCore.SMcScanPolygonGeometry;
    EditMode: MapCore.IMcEditMode;
    SpatialQueryParams: MapCore.IMcSpatialQueries.SQueryParams;
    CompletelyInside: boolean;
    ScanCoordSys: MapCore.EMcPointCoordSystem;
    prevCallback: any;
    viewportData: ViewportData;

    constructor(coordSys: MapCore.EMcPointCoordSystem, EditMode: MapCore.IMcEditMode, CompletelyInsideOnly?: boolean, Tolerance?: Number, SQParams?: MapCore.IMcSpatialQueries.SQueryParams) {
        this.CompletelyInside = CompletelyInsideOnly;
        this.SpatialQueryParams = SQParams;
        this.EditMode = EditMode;
        this.ScanCoordSys = coordSys;
        this.prevCallback = new this.IMcEditModeCallbacks();
    }
    public StartPolyScan(viewportData: ViewportData) {
        this.viewportData = viewportData;
        let activeOverlay = objectWorldService.findActiveOverlayByMcViewport(viewportData.viewport)
        let locationPoints: MapCore.SMcVector3D[] = [];
        let ObjSchemeItem: MapCore.IMcPolygonItem = null;
        runMapCoreSafely(() => {
            ObjSchemeItem = MapCore.IMcPolygonItem.Create((MapCore.IMcObjectSchemeItem.EItemSubTypeFlags.EISTF_SCREEN as any).value,
                MapCore.IMcLineBasedItem.ELineStyle.ELS_SOLID,
                MapCore.bcBlackOpaque,
                2,
                null,
                new MapCore.SMcFVector2D(0, -1),
                1,
                MapCore.IMcLineBasedItem.EFillStyle.EFS_SOLID,
                new MapCore.SMcBColor(0, 255, 100, 100)
            );
        }, "scanPolygonGeometry.StartPolyScan => IMcPolygonItem.Create", true)

        let obj: MapCore.IMcObject = null;
        runMapCoreSafely(() => {
            obj = MapCore.IMcObject.Create(activeOverlay,
                ObjSchemeItem,
                MapCore.EMcPointCoordSystem.EPCS_SCREEN,
                locationPoints,
                false);
        }, "scanPolygonGeometry.StartPolyScan => IMcObject.Create", true)
        //In order to prevent retrieval of the polygon in the scan 
        runMapCoreSafely(() => { ObjSchemeItem.SetDetectibility(false); }, "scanPolygonGeometry.StartPolyScan => IMcObjectSchemeItem.SetDetectibility", true)
        runMapCoreSafely(() => {
            ObjSchemeItem.SetDrawPriorityGroup(MapCore.IMcSymbolicItem.EDrawPriorityGroup.EDPG_TOP_MOST
                , 2
            );
        }, "scanPolygonGeometry.StartPolyScan => IMcSymbolicItem.SetDrawPriorityGroup", true)
        runMapCoreSafely(() => { ObjSchemeItem.SetDrawPriority(127); }, "scanPolygonGeometry.StartPolyScan => IMcSymbolicItem.SetDrawPriority", true)

        // let activeViewportID: Uint8Array;
        // activeViewportID[0] = viewportData.id;
        let viewportSelector: MapCore.IMcConditionalSelector = null;
        runMapCoreSafely(() => {
            viewportSelector = MapCore.IMcViewportConditionalSelector.Create(viewportData.viewport.GetOverlayManager(),
                // MapCore.IMcViewportConditionalSelector.EViewportTypeFlags.EVT_ALL_VIEWPORTS,
                // MapCore.IMcViewportConditionalSelector.EViewportCoordinateSystem.EVCS_ALL_COORDINATE_SYSTEMS,
                // activeViewportID,
                // true
            );
        }, "scanPolygonGeometry.StartPolyScan => IMcViewportConditionalSelector.Create", true)
        runMapCoreSafely(() => {
            obj.SetConditionalSelector(MapCore.IMcConditionalSelector.EActionType.EAT_VISIBILITY,
                true,
                viewportSelector);
        }, "scanPolygonGeometry.StartPolyScan => IMcObject.SetConditionalSelector", true)
        runMapCoreSafely(() => { obj.SetDrawPriority(127); }, "scanPolygonGeometry.StartPolyScan => IMcObject.SetDrawPriority", true)
        runMapCoreSafely(() => { viewportData.editMode.StartInitObject(obj, ObjSchemeItem); }, "scanPolygonGeometry.StartPolyScan => IMcEditMode.StartInitObject", true)
        runMapCoreSafely(() => { this.prevCallback = this.EditMode.GetEventsCallback(); }, "scanPolygonGeometry.StartPolyScan => IMcEditMode.GetEventsCallback", true)
        runMapCoreSafely(() => { viewportData.editMode.SetEventsCallback(new this.IMcEditModeCallbacks()); }, "scanPolygonGeometry.StartPolyScan => IMcEditMode.SetEventsCallback", true)
    }
    IMcEditModeCallbacks: any = this.createIMcEditModeCallbacksClass(this);
    private createIMcEditModeCallbacksClass(data: ScanPolygonGeometry): MapCore.IMcEditMode.ICallback {
        return MapCore.IMcEditMode.ICallback.extend("IMcEditMode.ICallback",
            {
                // optional
                ExitAction: function () {
                    data.EditMode.SetEventsCallback(data.prevCallback);
                },

                // optional
                Release: function () {
                    // this.delete();
                },
                InitItemResults(pObject: MapCore.IMcObject, pItem: MapCore.IMcObjectSchemeItem, nExitCode: number) {
                    if (nExitCode != 0) {
                        let objScheme: MapCore.IMcObjectScheme = null;
                        runMapCoreSafely(() => { objScheme = pObject.GetScheme(); }, "scanPolygonGeometry.createIMcEditModeCallbacksClass => IMcObject.GetScheme", true)
                        let objSchemeNode: MapCore.IMcObjectSchemeNode[] = null;
                        runMapCoreSafely(() => { objSchemeNode = objScheme.GetNodes((MapCore.IMcObjectSchemeNode.ENodeKindFlags.ENKF_ANY_ITEM as any).value); }, "scanPolygonGeometry.createIMcEditModeCallbacksClass => IMcObjectScheme.GetNodes", true)
                        let Coords: MapCore.SMcVector3D[] = null;
                        runMapCoreSafely(() => { Coords = objSchemeNode[0].GetCoordinates(data.viewportData.viewport, data.ScanCoordSys, pObject); }, "scanPolygonGeometry.createIMcEditModeCallbacksClass => IMcObjectSchemeNode.GetCoordinates", true)
                        for (let idx = 0; idx < Coords.length; ++idx) { Coords[idx].z = 0; }
                        //Remove rectangle Item from the map
                        runMapCoreSafely(() => { pObject.Remove(); }, "scanPolygonGeometry.createIMcEditModeCallbacksClass => IMcObject.Remove", true)
                        // if (Coords.length == 2) {
                        let boxCoord: MapCore.SMcBox = new MapCore.SMcBox(Coords[0], Coords[1]);
                        data.ScanPolygonGeometry = new MapCore.SMcScanBoxGeometry(data.ScanCoordSys, boxCoord);
                        data.ScanPolygonGeometry.uGeometryType = MapCore.SMcScanBoxGeometry.GEOMETRY_TYPE;
                        let TargetFound: MapCore.IMcSpatialQueries.STargetFound[];
                        runMapCoreSafely(() => { TargetFound = data.viewportData.viewport.ScanInGeometry(data.ScanPolygonGeometry, data.CompletelyInside, data.SpatialQueryParams); }, "scanPolygonGeometry.createIMcEditModeCallbacksClass => IMcSpatialQueries.ScanInGeometry", true)
                        mapcoreData.TargetFound = TargetFound;
                        mapcoreData.whenTargetFoundFilled();
                        mapcoreData.ScanGeometry = data.ScanPolygonGeometry;
                    }
                }
            });
    }
}

export default ScanPolygonGeometry